lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -289,6 +294,106 @@ def filter_query(filters: list[str], query: Query) -> Query:
+            value: the current metadata key's value
+        """
+        if key in keys:
+            p = keys[key]
+        else:
+            p = {}
+            keys[key] = p
+        if isinstance(value, dict):
+            for k, v in value.items():
+                self.accumulate(p, k, v)
+
+    def keyspace(self, query: Query) -> JSONOBJECT:
+        """Aggregate the dataset metadata keyspace
+
+        Run the query we've compiled, but instead of returning Dataset proxies,
+        we only want the metadata key/value pairs we've selected.
+
+        NOTE: In general we expect every dataset to have some metadata; however
+        in the event one doesn't, the key and value returned by the joined
+        query will be None. We handle this by ignoring those values.
+
+        Args:
+            query: The basic filtered SQLAlchemy query object
+
+        Returns:
+            The aggregated keyspace JSON object
+        """
+        keys: JSONOBJECT = {"dataset": {c.name: {} for c in Dataset.__table__._columns}}
+        list = query.with_entities(Metadata.key, Metadata.value).all()
+        for k, v in list:
+            # "metalog" is a top-level key in the Metadata schema, but we
+            # report it as a sub-key of "dataset".
+            if k == Metadata.METALOG:
+                self.accumulate(keys["dataset"], k, v)
+            elif k:
+                self.accumulate(keys, k, v)
+        return keys
+
+    def datasets(self, request: Request, json: JSONOBJECT, query: Query) -> JSONOBJECT:
+        """Gather and paginate the selected datasets
+
+        Run the query we've compiled, with pagination limits applied; collect
+        results into a list of JSON objects including selected metadata keys.
+
+        Args:
+            request: The HTTP Request object
+            json: The JSON query parameters
+            query: The basic filtered SQLAlchemy query object
+
+        Returns:
+            The paginated dataset listing
+        """
+        try:
+            datasets, paginated_result = self.get_paginated_obj(
+                query=query, json=json, url=request.url
+            )
+        except (AttributeError, ProgrammingError, StatementError) as e:
+            raise APIInternalError(
+                f"Constructed SQL for {json} isn't executable"
+            ) from e
+        except Exception as e:
+            raise APIInternalError(f"Unexpected SQL exception: {e}") from e
+
+        keys = json.get("metadata")
+
+        response = []
+        for dataset in datasets:
+            d = {
+                "name": dataset.name,
+                "resource_id": dataset.resource_id,
+            }
+            try:
+                d["metadata"] = self._get_dataset_metadata(dataset, keys)
+            except MetadataError as e:
+                current_app.logger.warning(
+                    "Error getting metadata {} for dataset {}: {}", keys, dataset, e
+                )
```

Comment:
Yeah; I didn't think about this when I moved the block, but I *have* been saying we should be removing warning and even info logs from APIs. The central question here would be whether the `MetadataError` here could possibly result from something internal like a SQL connection error that might not have already been logged in deeper levels. If it's strictly user errors, we don't want it. I guess I'll skim through the call tree ...

Reply:
I admit I punted on this in the 2nd commit. I thought about it. There are "cases". A SQL error will be logged in `Metadata.get`, so that's fine. An invalid key path should be diagnosed in parameter validation. I *think* the main case we might hit here would be an inconsistent keyspace reference; e.g., where you have `global.data: <scalar>` and you try to get `?metadata=global.data.foo`. In other metadata contexts (e.g., `GET/PUT /datasets/{id}/metadata`) I'm returning an `error` field in the JSON response providing messages like that. I wasn't sure I wanted to do that here, nor did I want to completely remove all record of the error. So I hesitated to just delete the log message and ended up pushing the commit without going back to think on it again. I think this is something I'd rather not worry about right now, so decide whether you want to keep pushing... the cliff edge is crumbly, so it might not take much. 😆

Reply:
> I think this is something I'd rather not worry about right now

OK.  Given that you said,

> I have been saying we should be removing warning and even info logs from APIs.

we can remove this when we do that.  😉